### ISSUE TITLE
ZIP file support

### ISSUE THREAD

**Issue by JuniorJPDJ** [mxmlnkn/ratarmount#64]
Hi!
Amazing work with tars ;D
It would be amazing to see support for ZIP files with all this features like recursive mounting.
ZIP file is AFAIK the most used archive format.

It shouldn't be very hard as zip is already supported by python itself.
There's [something like this](https://github.com/openscopeproject/ZipROFS/issues) but without recursive mounts, but mixing archive types in recursive mounting would be amazing feature.

If you think it's out of scope - just close with short note ;)

**Comment by mxmlnkn** [comment]
I think zip is only the most used file-format on windows? And I'm not sure how well ratarmount works on Windows, at least it was not tested.

The decision about the scope is hard. I do see the usefulness and it would fit if ra**tar**mount went more into the direction of **archive**mount. But then, it would be even more in close competition with archivemount and I'm not sure whether ratarmount is actually better for all cases. In your case, recursive mounting would be the discerning feature, which archivemount does not have out of the box.

One internal problem is that zip files do not map at all to the index file. It would require some arduous abstraction refactoring of the file (metadata) access and index database. However, maybe it might be an alleviating factor that zip might support seeking out of the box, so a sidecar file for seeking would maybe not be necessary.

As for recursive mounting, there are some limitations currently because of that index file and the missing abstraction! Currently, it only works for uncompressed TAR files contained in other tar files. Having a tar.bz2 inside another .tar or .tar.bz2 will not mount that internal .tar.bz2 recursively! I might open an issue to add support for this and that could pave the way for zip and other archive formats.

For now, I'll keep it open, it doesn't bother me.

**Comment by JuniorJPDJ** [comment]
Anyway: my usecase would be mounting all zips and tars from one directory with dynamic updating and without archive-in-archive recursive case (something like #56).
ZIP files have something like index inside, at end of archive.
There's what I'm using now: https://github.com/openscopeproject/ZipROFS

**Comment by mxmlnkn** [comment]
> One internal problem is that zip files do not map at all to the index file. It would require some arduous abstraction refactoring of the file (metadata) access and index database. However, maybe it might be an alleviating factor that zip might support seeking out of the box, so a sidecar file for seeking would maybe not be necessary.

Thinking about it again, this might be easier than thought because I already have a kind of MountSource interface albeit not specified explicitly. But, from the perspective of the TarMount class, `FolderMountSource` and `SQLiteIndexedTar` can be used almost interchangeably. The first step would be to define an abstract base class `MountSource` and in the next step a `ZipMountSource` would have to be implemented and used when a zip file has been recognized.

**Comment by JuniorJPDJ** [comment]
Seems super-cool :D

**Comment by mxmlnkn** [comment]
Ok, so I hacked something together in 2c212a05 in the develop branch. You can try it out with:

```bash
pip install git+https://github.com/mxmlnkn/ratarmount.git@develop#egginfo=ratarmount
```

It should be able to mount .zip files directly and when bind mounting (with `--recursive`) a folder containing a zip.

The next two steps would be to split the TarMount class into a generic `UnionMountSource` and a much simpler `TarFuseMount` class.

After that much needed refactor, I may even look into recursive mounting over different archive compressions, but the first version will most likely be a memory hog as I intend to make all `MountSource` derived classes offer an `open` method returning a file object, from which I can then construct another mount source.

**Comment by mxmlnkn** [comment]
There seems to be a larger conceptual issue with recursive mounting:

 - Most of the compression backends not written in pure Python need either a path or a file descriptor to access the file and won't work with arbitrary Python-level file objects.
 - Trying to open a file inside the mount point from inside a FUSE callback to get an OS-level file descriptor is a very bad idea! This will make the process hang in uninterruptible sleep because it is basically waiting for itself, at least in single-threaded FUSE mode. Maybe I can make it work with FUSE multi-thread mode but that might introduce a whole other plethora of bugs.

Fun fact, I can't find any way to close that botched ratarmount process now. Neither `kill -9`, no killing the parent process nor `fusermount -u` nor anything works. Now I basically have a block hole mountpoint, which will hang any process trying to access it.  Combine this with a loop like `while true; ls <mount point> &; done` and you can hit the process limit with only unkillable processes...

**Comment by JuniorJPDJ** [comment]
> Maybe it is possible with the Python C API to use python file objects directly but that would require adjusting all backends, currently, three indexed_bzip2, indexed_zstd, indexed_gzip and probably others if I wanted to add more like rar and 7z. 

This seems to be the most correct idea IMO but it will probably need loads of work.
Some people wanted to use your indexed backends as python library anyway so maybe it will solve 2 problems at once?

> zipfile actually seems to work for some reason, I guess it is written in pure Python but that might even depend on the compression used inside the zip.

zipfile is pure python and using available compression methods which could be pure python or C extensions.

**Comment by mxmlnkn** [comment]
> > Maybe it is possible with the Python C API to use python file objects directly but that would require adjusting all backends, currently, three indexed_bzip2, indexed_zstd, indexed_gzip and probably others if I wanted to add more like rar and 7z.
> 
> This seems to be the most correct idea IMO but it will probably need loads of work.
> Some people wanted to use your indexed backends as python library anyway so maybe it will solve 2 problems at once?

It actually solved a myriad of problems to abstract the file reading stuff away into the already existing but not fully utilized FileReader interface. It was a major refactor but it was very rewarding as many things fell in place. Here is a list of them from the refactor commit:

 - This fixes weird file read issues on macOS.
 - By taking a `FileReader`, the multitude of constructors for `BitReader`, `BZ2Reader`, `ParallelBZ2Reader`, `BlockFinder`, `BitStringFinder`, and `ParallelBitStringFinder` can be reduced to ideally one.
 - The complexities with refilling the buffer from either a file descriptor, or file, or buffer, in `BitReader` are moved to separate
   classes.
 - The complexities with copying file readers while ensuring that the file positions of the underlying files can be independently changed by multiple threads is moved into the `FileReader::clone` method.
 - The new `SharedFileReader` class avoids depending on the `/dev/fd/<fd>` trick to reopen a file with an independent file position. It opens the file once and then serializes access from multiple threads to it.
 - Also note that on macOS, opening `/dev/fd/<fd>` might not start with the file position at 0, which might indicate shared a file position!
 - Both, the `FileReader` as constructor argument and the `SharedFileReader` are prerequisites for adding Python file object handling support!

Now, I'll have to see how well it fares with the ratarmount refactor and recursive mounting capabilities.

**Comment by mxmlnkn** [comment]
Fixed on master and I will release v0.9.0 today.

The AutoMountLayer can now recursively mount many archives. You can now mount a zip inside a rar inside a .tar.bz2 inside a .tar.gz and so on.

I also added `--password` and `--password-file` options. The latter is a list of passwords, one per line, which will be tried out in order for encrypted archives. This way you can recursively mount many different archives with different passwords.
